master/internal/resourcemanagers/agent/agent_state.go

```diff
@@ -0,0 +1,375 @@
+		return nil, errors.New("not enough devices")
+	}
+
+	for _, d := range devices {
+		a.Devices[d] = &cid
+	}
+
+	return devices, nil
+}
+
+// DeallocateContainer deallocates containers.
+func (a *AgentState) DeallocateContainer(id cproto.ID) {
+	delete(a.ZeroSlotContainers, id)
+	for d, cid := range a.Devices {
+		if cid != nil && *cid == id {
+			a.Devices[d] = nil
+		}
+	}
+}
+
+// DeepCopy returns a copy of agentState for scheduler internals.
+func (a *AgentState) DeepCopy() *AgentState {
+	copiedAgent := &AgentState{
+		Handler:               a.Handler,
+		Label:                 a.Label,
+		Devices:               make(map[device.Device]*cproto.ID),
+		ZeroSlotContainers:    make(map[cproto.ID]bool),
+		maxZeroSlotContainers: a.maxZeroSlotContainers,
+		enabled:               a.enabled,
+		draining:              a.draining,
+		// TODO(ilia): Deepcopy of `slotStates` may be necessary one day.
+		slotStates: a.slotStates,
+	}
+
+	for originalDevice, id := range a.Devices {
+		copiedDevice := device.Device{
+			ID:    originalDevice.ID,
+			Brand: originalDevice.Brand,
+			UUID:  originalDevice.UUID,
+			Type:  originalDevice.Type,
+		}
+		copiedAgent.Devices[copiedDevice] = id
+	}
+
+	for originalKey, originalValue := range a.ZeroSlotContainers {
+		copiedAgent.ZeroSlotContainers[originalKey] = originalValue
+	}
+
+	return copiedAgent
+}
+
+// Enable enables the agent.
+func (a *AgentState) Enable(ctx *actor.Context) {
+	ctx.Log().Infof("enabling agent: %s", a.string())
+	a.enabled = true
+	a.draining = false
+}
+
+// Disable disables or drains the agent.
+func (a *AgentState) Disable(ctx *actor.Context, drain bool) {
+	drainStr := "disabling"
+	if drain {
+		drainStr = "draining"
+	}
+	ctx.Log().Infof("%s agent: %s", drainStr, a.string())
+	a.draining = drain
+	a.enabled = false
+}
+
+func (a *AgentState) addDevice(ctx *actor.Context, device device.Device, containerID *cproto.ID) {
+	ctx.Log().Infof("adding device: %s on %s", device.String(), a.string())
+	a.Devices[device] = containerID
+}
+
+func (a *AgentState) removeDevice(ctx *actor.Context, device device.Device) {
+	ctx.Log().Infof("removing device: %s (%s)", device.String(), a.string())
+	delete(a.Devices, device)
+}
+
+// agentStarted initializes slots from AgentStarted.Devices.
+func (a *AgentState) agentStarted(ctx *actor.Context, agentStarted *aproto.AgentStarted) {
+	msg := agentStarted
+	for _, d := range msg.Devices {
+		enabled := slotEnabled{
+			agentEnabled: true,
+			userEnabled:  true,
+		}
+		a.slotStates[d.ID] = &slot{enabled: enabled, device: d}
+		a.updateSlotDeviceView(ctx, d.ID)
+	}
+}
+
+func (a *AgentState) containerStateChanged(ctx *actor.Context, msg aproto.ContainerStateChanged) {
+	for _, d := range msg.Container.Devices {
+		s, ok := a.slotStates[d.ID]
+		if !ok {
+			ctx.Log().Warnf("bad containerStateChanged on device: %d (%s)", d.ID, a.string())
+			continue
+		}
+
+		s.container = &msg.Container
+		if msg.Container.State == cproto.Terminated {
```

Comment:
should we also, finally, `delete(a.containers, msg.Container.ID)`. i dont see that map being deleted from

Reply:
that happens in `agent.containerStateChanged` (which calls this method).